core/src/main/java/hudson/model/UpdateCenter.java

```diff
@@ -1177,8 +1177,19 @@ public File download(DownloadJob job, URL src) throws IOException {
          * Connects to the given URL for downloading the binary. Useful for tweaking
          * how the connection gets established.
          */
-        protected URLConnection connect(DownloadJob job, URL src) throws IOException {
-            return ProxyConfiguration.open(src);
+        protected URLConnection connect(DownloadJob job, URL src) throws IOException { 
+            URLConnection conn = ProxyConfiguration.open(src);
+            if(conn instanceof HttpURLConnection) {
+                HttpURLConnection httpConn = (HttpURLConnection) conn;
+                if (httpConn.getResponseCode() == HttpURLConnection.HTTP_MOVED_TEMP 
+                    || httpConn.getResponseCode() == HttpURLConnection.HTTP_MOVED_PERM) {
+                    // 302, 301
+                    // TODO maybe need resolve recursive redirect
+                    conn = connect(job, new URL(httpConn.getHeaderField("location")));
+                    IOUtils.closeQuietly(httpConn.getInputStream());
```

Comment:
It won't be closed on exceptional paths.
Please also use Util.closeAndLogFailures() instead, because we want to report the issue in logs at least